File: apps/android/shared/src/main/java/dev/agentdeck/shared/BridgeClient.kt
package dev.agentdeck.shared

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.withTimeoutOrNull
import kotlinx.coroutines.isActive
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.Call
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.net.InetAddress
import java.net.Proxy
import java.net.ProxySelector
import java.net.SocketAddress
import java.net.URI
import java.util.TimeZone
import java.util.concurrent.TimeUnit

/**
 * Connects to addresses on this device's own network directly, and everything
 * else however the system says.
 *
 * Wear OS installs a system-wide SOCKS proxy that tunnels app traffic to the
 * phone over Bluetooth — including, uselessly, traffic for the LAN the watch
 * itself sits on, where the tunnel just times out. The shell's curl reached
 * the bridge while the app reported it unreachable, because only the app was
 * being routed through the proxy. A LAN-literal bridge address never benefits
 * from a tunnel to another device, so it is dialed directly.
 */
private class LanDirectProxySelector : ProxySelector() {
    private val system = getDefault()

    override fun select(uri: URI?): List<Proxy> {
        val host = uri?.host ?: return listOf(Proxy.NO_PROXY)
        val literal = runCatching {
            // Only literals: a hostname would trigger DNS on this thread.
            if (host.matches(Regex("[0-9.]+")) || host.contains(":")) InetAddress.getByName(host) else null
        }.getOrNull()
        if (literal != null && (literal.isSiteLocalAddress || literal.isLoopbackAddress || literal.isLinkLocalAddress)) {
            return listOf(Proxy.NO_PROXY)
        }
        return system?.select(uri) ?: listOf(Proxy.NO_PROXY)
    }

    override fun connectFailed(uri: URI?, address: SocketAddress?, failure: IOException?) {
        system?.connectFailed(uri, address, failure)
    }
}

/**
 * The bridge declined to hand a message to a session blocked on a person.
 *
 * The message carries the bridge's own sentence about what is pending, so the
 * composer can point at the approval or question instead of reporting a code.
 */
class AgentBlockedException(detail: String) : Exception(detail)

class BridgeClient(
    baseUrl: String,
    private var token: String = "",
    /**
     * Sockets bound to a specific network, for a device whose default network
     * cannot reach the bridge — the watch's default is the Bluetooth companion
     * tunnel, and the bridge sits on the Wi-Fi beside it.
     */
    socketFactory: javax.net.SocketFactory? = null,
) {
    private val json = Json { ignoreUnknownKeys = true }
    private val http = OkHttpClient.Builder()
        .apply { socketFactory?.let { socketFactory(it) } }
        .proxySelector(LanDirectProxySelector())
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()
    private val streamHttp = http.newBuilder()
        .readTimeout(0, TimeUnit.SECONDS)
        .build()
    @Volatile private var activeStreamCall: Call? = null

    var baseUrl: String = normalizeUrl(baseUrl)
        private set

    fun configure(url: String, newToken: String) {
        baseUrl = normalizeUrl(url)
        token = newToken.trim()
    }

    suspend fun snapshot(): BridgeSnapshot = withContext(Dispatchers.IO) {
        val request = requestBuilder("$baseUrl/bridge/v1/snapshot").get().build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("Bridge returned ${response.code}")
            json.decodeFromString<BridgeSnapshot>(response.body.string())
        }
    }

    /**
     * Every file change a session produced. The live snapshot carries only a rolling window of
     * events, so a long session's earlier edits are not in it — the Changes tab loads them here.
     */
    suspend fun changes(agentId: String): List<AgentEvent> = withContext(Dispatchers.IO) {
        val request = requestBuilder("$baseUrl/bridge/v1/agents/$agentId/changes").get().build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("Bridge returned ${response.code}")
            json.decodeFromString<AgentChanges>(response.body.string()).changes
        }
    }

    /**
     * A session's retained history. The live snapshot only carries a rolling window sized for
     * cards, so on a busy session the conversation, reasoning and terminal output all age out of
     * it — the session view reads them here instead.
     */
    /**
     * A session's retained history. `limit` asks the bridge for only the most
     * recent events, for a caller that cannot afford the whole thing.
     */
    suspend fun history(agentId: String, limit: Int? = null): List<AgentEvent> = withContext(Dispatchers.IO) {
        val suffix = limit?.let { "?limit=$it" }.orEmpty()
        val request = requestBuilder("$baseUrl/bridge/v1/agents/$agentId/history$suffix").get().build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("Bridge returned ${response.code}")
            json.decodeFromString<AgentHistory>(response.body.string()).events
        }
    }

    /** What this session can be asked to run by name, for the composer's "/" picker. */
    suspend fun slashCommands(agentId: String): List<SlashCommand> = withContext(Dispatchers.IO) {
        val request = requestBuilder("$baseUrl/bridge/v1/agents/$agentId/slash-commands").get().build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("Bridge returned ${response.code}")
            json.decodeFromString<SlashCommandCatalog>(response.body.string()).commands
        }
    }

    suspend fun streamSnapshots(onSnapshot: (BridgeSnapshot) -> Unit) = withContext(Dispatchers.IO) {
        val request = requestBuilder("$baseUrl/bridge/v1/events")
            .header("Accept", "text/event-stream")
            .get()
            .build()
        val call = streamHttp.newCall(request)
        activeStreamCall = call
        call.execute().use { response ->
            if (!response.isSuccessful) error("Event stream returned ${response.code}")
            val source = response.body.source()
            var eventName = ""
            var latest: BridgeSnapshot? = null
            val data = StringBuilder()
            while (currentCoroutineContext().isActive && !source.exhausted()) {
                val line = source.readUtf8Line() ?: break
                when {
                    line.startsWith("event:") -> eventName = line.substringAfter(':').trim()
                    line.startsWith("data:") -> data.append(line.substringAfter(':').trim())
                    line.isEmpty() -> {
                        if (data.isNotEmpty()) when (eventName) {
                            "snapshot" -> {
                                val snapshot = json.decodeFromString<BridgeSnapshot>(data.toString())
                                latest = snapshot
                                onSnapshot(snapshot)
                            }
                            // A patch only makes sense against the snapshot it was computed from;
                            // without one the connection is not yet synced, so wait for the full send.
                            "patch" -> latest?.let { previous ->
                                val merged = previous.applyPatch(json.decodeFromString<BridgeSnapshotPatch>(data.toString()))
                                latest = merged
                                onSnapshot(merged)
                            }
                        }
                        eventName = ""
                        data.clear()
                    }
                }
            }
        }
        activeStreamCall = null
    }

    fun cancelStream() {
        activeStreamCall?.cancel()
        activeStreamCall = null
    }

    /**
     * Tells the bridge a person just looked at this session, so every other
     * surface can drop its badge. Fire-and-forget at call sites: the local
     * mark has already applied by the time this is sent, and a failure costs
     * only the echo to the other surfaces.
     */
    suspend fun markSeen(agentId: String) = withContext(Dispatchers.IO) {
        val body = ByteArray(0).toRequestBody("application/json".toMediaType())
        val request = requestBuilder("$baseUrl/bridge/v1/agents/$agentId/seen").post(body).build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("Seen returned ${response.code}")
        }
    }

    suspend fun analytics(range: String, project: String? = null): AnalyticsSnapshot = withContext(Dispatchers.IO) {
        val url = "$baseUrl/bridge/v1/analytics".toHttpUrl().newBuilder()
            .addQueryParameter("range", range)
            .addQueryParameter("timeZone", TimeZone.getDefault().id)
            .apply { project?.let { addQueryParameter("project", it) } }
            .build()
        val request = requestBuilder(url.toString()).get().build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("Analytics returned ${response.code}")
            json.decodeFromString<AnalyticsSnapshot>(response.body.string())
        }
    }

    suspend fun pair(code: String, deviceName: String): PairedDevice = withContext(Dispatchers.IO) {
        val body = json.encodeToString(PairRequest(code, deviceName))
            .toRequestBody("application/json".toMediaType())
        val request = Request.Builder().url("$baseUrl/bridge/v1/pair").post(body).build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                val message = runCatching { response.body.string() }.getOrDefault("")
                error(if (response.code == 401) "Pairing code is invalid or expired" else "Pairing failed (${response.code}): $message")
            }
            json.decodeFromString<PairedDevice>(response.body.string())
        }
    }

    suspend fun control(agentId: String, action: String, value: String? = null, commandId: String? = null, force: Boolean = false) = withContext(Dispatchers.IO) {
        val body = json.encodeToString(ControlRequest(action, value, commandId, force = if (force) true else null))
            .toRequestBody("application/json".toMediaType())
        val request = requestBuilder("$baseUrl/bridge/v1/agents/$agentId/control").post(body).build()
        http.newCall(request).execute().use { response ->
            // A blocked session is a refusal with a reason, not a failure: the
            // bridge is saying "answer the pending request first, or say force".
            // It gets its own exception so the composer can offer exactly that,
            // while every other status keeps reading as the failure it is.
            if (response.code == 409) {
                val refusal = runCatching { json.decodeFromString<ControlRefusal>(response.body.string()) }.getOrNull()
                if (refusal?.error == "agent_blocked") {
                    throw AgentBlockedException(refusal.detail ?: "This agent is waiting on an approval or question.")
                }
            }
            if (!response.isSuccessful) error("Command failed (${response.code})")
        }
    }

    suspend fun answerQuestion(agentId: String, requestId: String, question: String, answer: String) = withContext(Dispatchers.IO) {
        val body = json.encodeToString(ManagedResolutionRequest("answered", mapOf(question to answer)))
            .toRequestBody("application/json".toMediaType())
        // Generic route: bridge-hosted sessions are answered in process, hook sessions durably.
        val request = requestBuilder("$baseUrl/bridge/v1/agents/$agentId/requests/$requestId/resolve").post(body).build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("Answer failed (${response.code})")
        }
    }

    suspend fun commandReceipt(commandId: String): CommandReceipt = withContext(Dispatchers.IO) {
        val request = requestBuilder("$baseUrl/bridge/v1/commands/$commandId/receipt").build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("Receipt unavailable (${response.code})")
            json.decodeFromString<CommandReceipt>(response.body.string())
        }
    }

    /** Which runtimes the bridge can host and run itself. */
    suspend fun managedRuntimes(): List<ManagedRuntime> = withContext(Dispatchers.IO) {
        val request = requestBuilder("$baseUrl/bridge/v1/managed/runtimes").get().build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("Managed runtimes unavailable (${response.code})")
            json.decodeFromString<ManagedRuntimes>(response.body.string()).runtimes
        }
    }

    /** Starts a bridge-hosted session. `cwd` must be absolute on the bridge's machine. */
    suspend fun startManagedSession(request: ManagedSessionRequest): StartedManagedSession = withContext(Dispatchers.IO) {
        val body = json.encodeToString(request).toRequestBody("application/json".toMediaType())
        val httpRequest = requestBuilder("$baseUrl/bridge/v1/managed/claude/sessions").post(body).build()
        http.newCall(httpRequest).execute().use { response ->
            if (!response.isSuccessful) {
                val message = runCatching { response.body.string() }.getOrDefault("")
                error("Could not start session (${response.code}): $message")
            }
            json.decodeFromString<StartedManagedSession>(response.body.string())
        }
    }

    private fun requestBuilder(url: String): Request.Builder = Request.Builder().url(url).apply {
        if (token.isNotBlank()) header("Authorization", "Bearer $token")
    }

    private fun normalizeUrl(value: String): String {
        val trimmed = value.trim().trimEnd('/')
        return when {
            trimmed.startsWith("http://") || trimmed.startsWith("https://") -> trimmed
            else -> "http://$trimmed"
        }
    }
}

class AgentRepository(private val client: BridgeClient) {
    private val _state = MutableStateFlow<BridgeState>(BridgeState.Loading)
    private var lastSequence = -1L
    private val wakeups = Channel<Unit>(Channel.CONFLATED)
    private val _connection = MutableStateFlow(ConnectionStatus("connecting"))
    val state: StateFlow<BridgeState> = _state.asStateFlow()
    val connection: StateFlow<ConnectionStatus> = _connection.asStateFlow()

    suspend fun stream(reconnectDelayMs: Long = 1_500, maxReconnectDelayMs: Long = 16_000) {
        var retryDelay = reconnectDelayMs
        var attempt = 0
        while (kotlin.coroutines.coroutineContext.isActive) {
            attempt += 1
            _connection.value = ConnectionStatus(if (lastSequence < 0) "connecting" else "reconnecting", attempt)
            val previous = currentSnapshot()
            var synchronized = false
            runCatching {
                client.streamSnapshots { snapshot ->
                    if (ConnectionPolicy.shouldApply(lastSequence, snapshot.sequence)) {
                        lastSequence = snapshot.sequence
                        synchronized = true
                        retryDelay = reconnectDelayMs
                        _connection.value = ConnectionStatus("connected", attempt)
                        _state.value = BridgeState.Ready(snapshot)
                    }
                }
            }.onFailure {
                val message = it.message ?: "Live connection lost"
                _state.value = BridgeState.Failed(message, previous)
                _connection.value = ConnectionStatus(if (ConnectionPolicy.isBlocked(message)) "blocked" else "backoff", attempt, if (ConnectionPolicy.isBlocked(message)) null else System.currentTimeMillis() + retryDelay, message)
            }
            if (_connection.value.phase == "blocked") wakeups.receive()
            else withTimeoutOrNull(retryDelay) { wakeups.receive() }
            if (!synchronized) {
                retryDelay = ConnectionPolicy.retryDelay(reconnectDelayMs, attempt + 1, maxReconnectDelayMs)
            } else {
                attempt = 0
            }
        }
    }

    private fun currentSnapshot() = when (val current = _state.value) {
        is BridgeState.Ready -> current.snapshot
        is BridgeState.Failed -> current.previous
        BridgeState.Loading -> null
    }

    suspend fun refresh() {
        val previous = currentSnapshot()
        runCatching { client.snapshot() }
            .onSuccess {
                if (ConnectionPolicy.shouldApply(lastSequence, it.sequence)) {
                    lastSequence = it.sequence
                    _state.value = BridgeState.Ready(it)
                }
            }
            .onFailure { _state.value = BridgeState.Failed(it.message ?: "Bridge unavailable", previous) }
    }

    suspend fun control(agentId: String, action: String, value: String? = null, commandId: String? = null, force: Boolean = false) {
        client.control(agentId, action, value, commandId, force)
        refresh()
    }

    suspend fun changes(agentId: String): List<AgentEvent> = client.changes(agentId)

    suspend fun history(agentId: String, limit: Int? = null): List<AgentEvent> =
        client.history(agentId, limit)

    suspend fun slashCommands(agentId: String): List<SlashCommand> = client.slashCommands(agentId)

    suspend fun answerQuestion(agentId: String, requestId: String, question: String, answer: String) {
        client.answerQuestion(agentId, requestId, question, answer)
        refresh()
    }

    suspend fun managedRuntimes(): List<ManagedRuntime> = client.managedRuntimes()

    suspend fun startManagedSession(request: ManagedSessionRequest): StartedManagedSession {
        val session = client.startManagedSession(request)
        // The new session arrives through the live stream like any other, but
        // a refresh means the deck shows it now rather than on the next patch.
        refresh()
        return session
    }

    fun wake() {
        client.cancelStream()
        wakeups.trySend(Unit)
    }

    fun configure(url: String, token: String) {
        client.configure(url, token)
        wake()
    }
}

File: apps/ios/AgentDeck/Net/BridgeClient.swift
import Foundation

/// Everything this app asks of the bridge, over `URLSession` and `Codable`.
///
/// Endpoints mirror `apps/android/shared/.../BridgeClient.kt` exactly — the
/// bridge is one contract, and a surface that invents its own paths is a
/// surface that breaks on a bridge upgrade nobody told it about.
actor BridgeClient {
    private(set) var baseURL: String
    private var token: String

    private let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        return decoder
    }()

    private let session: URLSession = {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 10
        configuration.waitsForConnectivity = false
        return URLSession(configuration: configuration)
    }()

    /// A stream has no read timeout: a quiet deck is not a broken one. The
    /// bridge sends a ping every 15 seconds, which is what actually proves the
    /// connection is alive.
    private let streamSession: URLSession = {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 3_600
        configuration.timeoutIntervalForResource = 86_400
        return URLSession(configuration: configuration)
    }()

    init(baseURL: String = "", token: String = "") {
        self.baseURL = BridgeClient.normalize(baseURL)
        self.token = token.trimmed
    }

    func configure(baseURL: String, token: String) {
        self.baseURL = BridgeClient.normalize(baseURL)
        self.token = token.trimmed
    }

    // MARK: - Reads

    func snapshot() async throws -> BridgeSnapshot {
        try await get("/bridge/v1/snapshot")
    }

    /// A session's retained history. The live snapshot carries only a rolling
    /// window sized for cards, so on a busy session the conversation ages out
    /// of it — the session view reads it here instead.
    func history(agentId: String, limit: Int? = nil) async throws -> [AgentEvent] {
        let suffix = limit.map { "?limit=\($0)" } ?? ""
        let response: AgentHistory = try await get("/bridge/v1/agents/\(escape(agentId))/history\(suffix)")
        return response.events
    }

    func changes(agentId: String) async throws -> [AgentEvent] {
        let response: AgentChanges = try await get("/bridge/v1/agents/\(escape(agentId))/changes")
        return response.changes
    }

    /// The models this session will answer as, asked of the runtime. A session
    /// the bridge does not host has no list — its model belongs to the runtime
    /// that owns its terminal — and answers 404, which is an empty list here so
    /// a surface can simply not offer the control.
    func models(agentId: String) async throws -> [RuntimeModel] {
        do {
            let response: RuntimeModels = try await get("/bridge/v1/agents/\(escape(agentId))/models")
            return response.models
        } catch BridgeError.http(404, _) {
            return []
        }
    }

    /// The message commands still queued for this session — the sender's to
    /// take back until the runtime collects them.
    func queuedMessages(agentId: String) async throws -> [QueuedCommand] {
        let response: QueuedCommands = try await get("/bridge/v1/agents/\(escape(agentId))/queued")
        return response.commands
    }

    /// Withdraws a queued message. A 404 means the runtime already collected
    /// it — the dock refreshes and the row disappears either way.
    func cancelQueued(agentId: String, commandId: String) async throws {
        do {
            _ = try await checked(
                try build("/bridge/v1/agents/\(escape(agentId))/queued/\(escape(commandId))", method: "DELETE"))
        } catch BridgeError.http(404, _) {
            // Already delivered; nothing left to withdraw.
        }
    }

    /// The `/` commands this runtime advertises. A runtime that publishes none
    /// answers with an empty catalog rather than a failure, so an empty list is
    /// an answer and not a swallowed error.
    func slashCommands(agentId: String) async throws -> [SlashCommand] {
        let catalog: SlashCommandCatalog = try await get("/bridge/v1/agents/\(escape(agentId))/slash-commands")
        return catalog.commands
    }

    /// Usage over a window. The time zone is the phone's, because "active days"
    /// and the heatmap's rows are calendar facts and a day boundary computed in
    /// UTC puts a late evening's work on tomorrow.
    func analytics(range: String, project: String?) async throws -> AnalyticsSnapshot {
        var components = URLComponents(string: baseURL + "/bridge/v1/analytics")
        var items = [
            URLQueryItem(name: "range", value: range),
            URLQueryItem(name: "timeZone", value: TimeZone.current.identifier),
        ]
        if let project { items.append(URLQueryItem(name: "project", value: project)) }
        components?.queryItems = items
        guard let url = components?.url else { throw BridgeError.unreachable("\(baseURL) is not a URL.") }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        if !token.isEmpty { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        return try decode(AnalyticsSnapshot.self, from: try await checked(request))
    }

    // MARK: - Writes

    /// Pairing is the one call made before a credential exists, so it carries
    /// no Authorization header at all.
    func pair(code: String, deviceName: String) async throws -> PairedDevice {
        let request = try build("/bridge/v1/pair", method: "POST", body: PairRequest(code: code, deviceName: deviceName), authorized: false)
        let (data, response) = try await perform(request, on: session)
        guard let http = response as? HTTPURLResponse else { throw BridgeError.malformed("No HTTP response") }
        guard (200 ..< 300).contains(http.statusCode) else {
            // The bridge answers 401 for a code that is wrong or has expired,
            // and 400 for one that is not six digits. Both are the code's
            // fault, and neither is worth a status number on screen.
            if http.statusCode == 401 { throw BridgeError.http(401, "That pairing code is invalid or expired.") }
            if http.statusCode == 400 { throw BridgeError.http(400, "A pairing code is six digits.") }
            throw BridgeError.http(http.statusCode, message(from: data))
        }
        return try decode(PairedDevice.self, from: data)
    }

    /// `force` is the explicit way past a 409: the bridge refuses to deliver a
    /// message to a session blocked on an approval or question unless the body
    /// says the sender saw that and meant it anyway.
    func control(agentId: String, action: String, value: String? = nil, commandId: String? = nil, force: Bool = false) async throws {
        let request = try build(
            "/bridge/v1/agents/\(escape(agentId))/control",
            method: "POST",
            body: ControlRequest(action: action, value: value, commandId: commandId, force: force ? true : nil)
        )
        _ = try await checked(request)
    }

    /// Answers a durable question. Only `answered` is a device's to send; the
    /// bridge refuses any other status from a paired credential with 403.
    func answer(agentId: String, requestId: String, question: String, answer: String) async throws {
        let request = try build(
            "/bridge/v1/agents/\(escape(agentId))/requests/\(escape(requestId))/resolve",
            method: "POST",
            body: ResolveRequest(value: [question: answer])
        )
        _ = try await checked(request)
    }

    /// Tells the bridge a person is looking at this session right now, so the
    /// other surfaces stop badging it. Read scope, no body — a view is a read,
    /// and the bridge stamps the moment itself.
    func markSeen(agentId: String) async throws {
        _ = try await checked(try build("/bridge/v1/agents/\(escape(agentId))/seen", method: "POST"))
    }

    /// Dismisses a session from the deck. Its history, usage and file changes
    /// are kept on the bridge — this declutters the live list, it does not
    /// erase what the session did. One still heartbeating simply reappears on
    /// its next beat.
    func dismiss(agentId: String) async throws {
        _ = try await checked(try build("/bridge/v1/agents/\(escape(agentId))", method: "DELETE"))
    }

    // MARK: - Managed sessions

    /// Which runtimes the bridge can host and run itself.
    func managedRuntimes() async throws -> [ManagedRuntime] {
        let response: ManagedRuntimes = try await get("/bridge/v1/managed/runtimes")
        return response.runtimes
    }

    /// Starts a bridge-hosted Claude session. `cwd` must be absolute on the
    /// bridge's machine, so the caller offers only paths a person knows it can
    /// reach — the bridge is the one that has to find the directory.
    func startManagedSession(request: ManagedSessionRequest) async throws -> StartedManagedSession {
        let urlRequest = try build("/bridge/v1/managed/claude/sessions", method: "POST", body: request)
        let data = try await checked(urlRequest)
        return try decode(StartedManagedSession.self, from: data)
    }

    // MARK: - Live stream

    /// One full snapshot, then patches. A patch only makes sense against the
    /// snapshot it was computed from, so one that arrives before the first full
    /// send is dropped rather than guessed at.
    func streamSnapshots(onSnapshot: @Sendable @escaping (BridgeSnapshot) -> Void) async throws {
        var request = try build("/bridge/v1/events", method: "GET")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.setValue("no-store", forHTTPHeaderField: "Cache-Control")

        let (bytes, response) = try await streamSession.bytes(for: request)
        guard let http = response as? HTTPURLResponse else { throw BridgeError.malformed("No HTTP response") }
        if http.statusCode == 401 || http.statusCode == 403 { throw BridgeError.unauthorized }
        guard (200 ..< 300).contains(http.statusCode) else { throw BridgeError.http(http.statusCode, nil) }

        var eventName = ""
        var data = ""
        var latest: BridgeSnapshot?

        /// One complete record. Reset by the caller either way, so a record the
        /// stream does not name is dropped rather than applied as the last one.
        func dispatch() throws {
            defer { eventName = ""; data = "" }
            guard !data.isEmpty, let payload = data.data(using: .utf8) else { return }
            switch eventName {
            case "snapshot":
                let snapshot = try decode(BridgeSnapshot.self, from: payload)
                latest = snapshot
                onSnapshot(snapshot)
            case "patch":
                // A patch only makes sense against the snapshot it was computed
                // from, so one that arrives before the first full send is
                // dropped rather than guessed at.
                guard let previous = latest else { return }
                let merged = previous.applying(try decode(BridgeSnapshotPatch.self, from: payload))
                latest = merged
                onSnapshot(merged)
            default:
                return
            }
        }

        // Raw bytes rather than `bytes.lines`.
        //
        // `AsyncLineSequence` does not yield empty lines — and the blank line
        // between records is the *only* thing that ends an SSE event. Parsing
        // this stream through it means every record runs into the next one and
        // nothing is ever dispatched, which is exactly what happened: the deck
        // showed whatever `/snapshot` last returned and no live update ever
        // reached it, while the connection sat there looking healthy.
        var line: [UInt8] = []
        line.reserveCapacity(8 * 1024)

        func consume(_ raw: [UInt8]) throws {
            guard let text = String(bytes: raw, encoding: .utf8) else { return }
            if text.hasPrefix("event:") {
                eventName = String(text.dropFirst("event:".count)).trimmed
            } else if text.hasPrefix("data:") {
                data += String(text.dropFirst("data:".count)).trimmed
            } else if text.isEmpty {
                try dispatch()
            }
            // `id:`, `retry:` and `:` comments are the stream's own bookkeeping
            // and carry nothing this client acts on.
        }

        for try await byte in bytes {
            guard byte == UInt8(ascii: "\n") else {
                line.append(byte)
                continue
            }
            // Tolerate CRLF framing from anything sitting in front of the bridge.
            if line.last == UInt8(ascii: "\r") { line.removeLast() }
            try consume(line)
            line.removeAll(keepingCapacity: true)
        }
    }

    // MARK: - Plumbing

    private func get<T: Decodable>(_ path: String) async throws -> T {
        let data = try await checked(try build(path, method: "GET"))
        return try decode(T.self, from: data)
    }

    private func checked(_ request: URLRequest) async throws -> Data {
        let (data, response) = try await perform(request, on: session)
        guard let http = response as? HTTPURLResponse else { throw BridgeError.malformed("No HTTP response") }
        if http.statusCode == 401 || http.statusCode == 403 { throw BridgeError.unauthorized }
        guard (200 ..< 300).contains(http.statusCode) else {
            // A 409 whose body names `agent_blocked` is the bridge protecting a
            // pending approval or question from a message that would race it.
            // A distinct error, because the composer answers it with a choice
            // rather than a failure line. Any other 409 stays a plain refusal.
            if http.statusCode == 409, let detail = blockedDetail(from: data) {
                throw BridgeError.agentBlocked(detail)
            }
            throw BridgeError.http(http.statusCode, message(from: data))
        }
        return data
    }

    /// The bridge's `{ "error": "agent_blocked", "detail": "…" }` body, or nil
    /// when the 409 is about something else.
    private func blockedDetail(from data: Data) -> String? {
        struct Blocked: Decodable {
            var error: String
            var detail: String?
        }
        guard let body = try? JSONDecoder().decode(Blocked.self, from: data), body.error == "agent_blocked" else { return nil }
        return body.detail ?? ""
    }

    private func perform(_ request: URLRequest, on session: URLSession) async throws -> (Data, URLResponse) {
        do {
            return try await session.data(for: request)
        } catch {
            throw BridgeError.from(error)
        }
    }

    private func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
        do {
            return try decoder.decode(type, from: data)
        } catch {
            throw BridgeError.malformed(String(describing: error))
        }
    }

    private func build(_ path: String, method: String, authorized: Bool = true) throws -> URLRequest {
        guard let url = URL(string: baseURL + path) else { throw BridgeError.unreachable("\(baseURL) is not a URL.") }
        var request = URLRequest(url: url)
        request.httpMethod = method
        if authorized, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    private func build(_ path: String, method: String, body: some Encodable, authorized: Bool = true) throws -> URLRequest {
        var request = try build(path, method: method, authorized: authorized)
        request.httpBody = try JSONEncoder().encode(body)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return request
    }

    private func escape(_ component: String) -> String {
        component.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? component
    }

    /// The bridge reports its own failures as `{"error": "..."}`; its wording
    /// is better than anything this app could invent for it.
    private func message(from data: Data) -> String? {
        struct Failure: Decodable { var error: String }
        return (try? JSONDecoder().decode(Failure.self, from: data))?.error
    }

    /// Fills in the scheme a person did not type.
    ///
    /// The two ways to reach a bridge look different, and the difference is
    /// exactly what says which scheme is meant. A direct one is an address and
    /// a port — `100.x.y.z:3000` — and speaks plain HTTP. A tailnet one is a
    /// MagicDNS name with no port, served over HTTPS by `tailscale serve`.
    ///
    /// Defaulting everything to `http://` sent a typed MagicDNS name to port
    /// 80, where nothing listens, and the app reported the bridge as out of
    /// range — which is true of port 80 and misleading about the bridge.
    static func normalize(_ value: String) -> String {
        var trimmed = value.trimmed
        while trimmed.hasSuffix("/") { trimmed.removeLast() }
        if trimmed.isEmpty { return "" }
        if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") { return trimmed }
        return "\(hasExplicitPort(trimmed) ? "http" : "https")://\(trimmed)"
    }

    /// Whether the authority carries a `:port`. Guards against an IPv6 literal,
    /// which is full of colons and has none of them meaning a port.
    private static func hasExplicitPort(_ authority: String) -> Bool {
        let host = authority.split(separator: "/", maxSplits: 1).first.map(String.init) ?? authority
        if host.hasPrefix("[") { return host.contains("]:") }
        let parts = host.split(separator: ":")
        guard parts.count == 2 else { return false }
        return parts[1].allSatisfy(\.isNumber)
    }
}
